File: haskell/Goal.hs
module Goal where

import Prelude hiding (all, any, either)
import State
import Term


_interleave :: [a] -> [a] -> [a]
_interleave []       ys = ys
_interleave (x : xs) ys = x : _interleave ys xs


type Goal a = State a -> [State a]

eq :: (Eq a) => Term a -> Term a -> Goal a
eq x y = stream . (unify x y)
  where
    stream :: Maybe a -> [a]
    stream (Just x) = [x]
    stream Nothing  = []

with :: [String] -> ([Term a] -> Goal a) -> Goal a
with names factory state = goal newState
  where
    (newState, vars) = mbind names state
    goal = factory vars

either :: Goal a -> Goal a -> Goal a
either x y state = _interleave (x state) (y state)

any :: [Goal a] -> Goal a
any (g : []) = g
any (g : gs) = either g $ any gs

both :: Goal a -> Goal a -> Goal a
both x y state = inEach y xstates
  where
    xstates           = x state
    inEach y []       = []
    inEach y (s : ss) = _interleave (y s) (inEach y ss)

all :: [Goal a] -> Goal a
all (g : []) = g
all (g : gs) = both g $ all gs

File: es6/lib/goal.js
import * as stream from './stream';

export default class Goal {
  constructor(func) {
    this._func = func;
  }

  pursue(state) {
    return this._func(state);
  }

  static equal(a, b) {
    return new Goal(state => {
      state = state.unify(a, b);

      return function*() {
        if (state) yield state;
      }();
    });
  }
  
  static bind(names, func) {
    return new Goal(state => {
      let [newState, vars] = state.createVars(...names),
          goal = func(...vars);

      return goal.pursue(newState);
    });
  }

  static either(a, b) {
    return new Goal(state => {
      return stream.interleave(a.pursue(state), b.pursue(state));
    });
  }

  static any(first, ...rest) {
    if (rest.length === 0)
      return first;
    else
      return Goal.either(first, Goal.any(...rest));
  }

  static both(a, b) {
    return new Goal(state => {
      return function*() {
        for (let state of a.pursue(state)) {
          for (let result of b.pursue(state))
            yield result;
        }
      }();
    });
  }

  static all(first, ...rest) {
    if (rest.length === 0)
      return first;
    else
      return Goal.both(first, Goal.all(...rest));
  }
}

Goal.eq  = Goal.equal;
Goal.or  = Goal.either;
Goal.and = Goal.both;
